isb_lib/opencontext_adapter.py

```diff
@@ -0,0 +1,175 @@
+        return _thing
+
+
+class OpenContextRecordIterator(isb_lib.core.IdentifierIterator):
+    def __init__(
+        self,
+        offset: int = 0,
+        max_entries: int = -1,
+        date_start: datetime.datetime = None,
+        date_end: datetime.datetime = None,
+        page_size: int = 100,
+    ):
+        super().__init__(
+            offset=offset,
+            max_entries=max_entries,
+            date_start=date_start,
+            date_end=date_end,
+            page_size=page_size,
+        )
+        self.url = OPENCONTEXT_API
+
+    def records_in_page(self):
+        L = get_logger()
+        L.debug("records_in_page")
+        headers = {"Accept": "application/json", "user-agent": "isamplesbot-3000/0.0.1"}
+        _page_size = self._page_size
+        params = {
+            "rows": _page_size,
+        }
+        more_work = True
+        num_records = 0
+        while more_work:
+            L.info("trying to hit %s", self.url)
+            response = requests.get(
+                self.url, params=params, headers=headers, timeout=HTTP_TIMEOUT
+            )
+
+            if response.status_code != 200:
+                L.error(
+                    "Unable to load records; status: %s; reason: %s",
+                    response.status_code,
+                    response.reason,
+                )
+                break
+            # L.debug("recordsInProject data: %s", response.text[:256])
+            data = response.json()
+            for record in data.get("oc-api:has-results", {}):
+                L.info("records_in_page Record id: %s", record.get("uri", None))
+                # print(json.dumps(record, indent=2))
+                # raise NotImplementedError
+                yield record
+                num_records += 1
+            self.url = data["next-json"]
+            if len(data.get("oc-api:has-results", {})) < _page_size:
+                more_work = False
+            elif 0 < self._max_entries <= num_records:
+                more_work = False
+            elif num_records == self._page_size:
+                more_work = False
+
+    def loadEntries(self):
+        self._cpage = []
+        self._page_offset = 0
+        counter = 0
+        for item in self.records_in_page():
+            self._cpage.append(item)
+            counter += 1
+            if 0 < self._max_entries < counter:
+                break
+        self._total_records = len(self._cpage)
+
+    def _getPage(self):
+        """Override this method to retrieve a page of entries from the service.
+
+        After completion, self._cpage contains the next page of entries or None if there
+        are no more pages available, and self._page_offset is set to the first entry (usually 0)
+
+        """
+        is_none = self._cpage is None
+        if is_none or self._page_offset >= len(self._cpage):
+            if not is_none:
+                # Work around the rate limiter
+                time.sleep(0.3)
```

Comment:
Not ideal, but it should ensure that we are executing roughly 3 requests/second.

Reply:
After setting the `user-agent` and running things again without the sleep, things are fine.  We take a bit of time to do the db inserts, so I don't think we need this after all.  It seems like the `user-agent` was the key.